File: Python/longest-valid-parentheses.py
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        # Time:  O(n)
        # Space: O(n)
        max_len = 0
        stack = [-1]
        for i, ch in enumerate(s):
            if ch == '(':
                stack.append(i)
            else:
                stack.pop()
                if not stack:
                    stack.append(i)
                else:
                    max_len = max(max_len, i - stack[-1])
        return max_len

File: C++/longest-valid-parentheses.cpp
#include <algorithm>
#include <stack>
#include <string>


class Solution {
public:
    int longestValidParentheses(std::string s) {
        // Time:  O(n)
        // Space: O(n)
        int maxLength = 0;
        std::stack<int> stk;
        stk.push(-1);

        for (size_t i = 0; i < s.size(); i++) {
            if (s[i] == '(') {
                stk.push(i);
            } else {
                stk.pop();
                if (stk.empty()) {
                    stk.push(i);
                } else {
                    maxLength = std::max<int>(maxLength, i - stk.top());
                }
            }
        }
        return maxLength;
    }
};